plugin-server/src/main/ingestion-queues/session-recording/blob-ingester/session-offset-high-water-mark.ts

```diff
@@ -0,0 +1,164 @@
+        const key = offsetHighWaterMarkKey(this.keyPrefix, tp)
+        try {
+            return await this.run(`read all offset high-water mark for ${key} `, async (client) => {
+                const redisValue = await client.zrange(key, 0, -1, 'WITHSCORES')
+                // redisValue is an array of [sessionId, offset, sessionId, offset, ...]
+                // we want to convert it to an object of { sessionId: offset, sessionId: offset, ... }
+                const highWaterMarks = redisValue.reduce(
+                    (acc: Record<string, number>, value: string, index: number) => {
+                        if (index % 2 === 0) {
+                            acc[value] = parseInt(redisValue[index + 1])
+                        }
+                        return acc
+                    },
+                    {}
+                )
+                if (highWaterMarks) {
+                    this.topicPartitionWaterMarks[`${tp.topic}-${tp.partition}`] = highWaterMarks
+                }
+                return highWaterMarks
+            })
+        } catch (error) {
+            status.error('🧨', 'WrittenOffsetCache failed to read high-water marks for partition', {
+                error: error.message,
+                key,
+                ...tp,
+            })
+            captureException(error, {
+                extra: {
+                    key,
+                },
+                tags: {
+                    ...tp,
+                },
+            })
+            return null
+        }
+    }
+
+    public async onCommit(tp: TopicPartition, offset: number): Promise<Record<string, number> | null> {
+        const key = offsetHighWaterMarkKey(this.keyPrefix, tp)
+        try {
+            return await this.run(`commit all below offset high-water mark for ${key} `, async (client) => {
+                const numberRemoved = await client.zremrangebyscore(key, '-Inf', offset)
+                status.info('📝', 'WrittenOffsetCache committed all below high-water mark for partition', {
+                    numberRemoved,
+                    ...tp,
+                    offset,
+                })
+                return await this.getAll(tp)
+            })
+        } catch (error) {
+            status.error('🧨', 'WrittenOffsetCache failed to commit high-water mark for partition', {
+                error: error.message,
+                key,
+                ...tp,
+            })
+            captureException(error, {
+                extra: {
+                    key,
+                },
+                tags: {
+                    ...tp,
+                },
+            })
+            return null
+        }
+    }
+
+    public async isBelowHighWaterMark(tp: TopicPartition, sessionId: string, offset: number): Promise<boolean> {
+        if (!this.topicPartitionWaterMarks[tp.partition]) {
+            const highWaterMarks = await this.getAll(tp)
```

Comment:
🤔 - if we process events in parallel there's a risk that this call happens _a lot_ until the promise settles. This is #NotGood

I think there are two things we could do:
1. Modify the consumer to not process in parallel [(this code)](https://github.com/PostHog/posthog/blob/master/plugin-server/src/main/ingestion-queues/session-recording/session-recordings-blob-consumer.ts#L227)
2. Change the `getAll` function to store a reference to the in progress promise so that we don't over-fetch.